src/widgets/notebook/NotebookEditor.tsx

```diff
@@ -0,0 +1,164 @@
+import { ActionIcon, createStyles, rem } from '@mantine/core';
+import { useDebouncedValue } from '@mantine/hooks';
+import { Link, RichTextEditor } from '@mantine/tiptap';
+import { IconArrowUp, IconEdit, IconEditOff } from '@tabler/icons-react';
+import { BubbleMenu, useEditor } from '@tiptap/react';
+import StarterKit from '@tiptap/starter-kit';
+import { useEffect, useRef, useState } from 'react';
+import { useConfigStore } from '~/config/store';
+import { useColorTheme } from '~/tools/color';
+import { api } from '~/utils/api';
+
+import { useEditModeStore } from '../../components/Dashboard/Views/useEditModeStore';
+import { useConfigContext } from '../../config/provider';
+import { WidgetLoading } from '../loading';
+import { INotebookWidget } from './NotebookWidgetTile';
+
+Link.configure({
+  openOnClick: true,
+});
+
+export function Editor({ widget }: { widget: INotebookWidget }) {
+  const [content, setContent] = useState(widget.properties);
+
+  const { enabled } = useEditModeStore();
+  const [isEditing, setIsEditing] = useState<boolean>(false);
+
+  const { config, name: configName } = useConfigContext();
+  const updateConfig = useConfigStore((x) => x.updateConfig);
+  const { primaryColor } = useColorTheme();
+
+  const { mutateAsync } = api.notebook.createOrUpdate.useMutation();
+
+  const [debounced] = useDebouncedValue(content, 500);
+
+  if (!config || !configName) return <WidgetLoading />;
+
+  const editor = useEditor({
+    extensions: [StarterKit, Link],
+    content: content.content,
+    editable: false,
+    onUpdate: (e) => {
+      setContent((previous) => ({
+        ...previous,
+        content: e.editor.getHTML(),
+      }));
+    },
+  });
+  useEffect(() => {
+    if (!editor) return;
+    editor.setEditable(isEditing);
+
+    updateConfig(
```

Comment:
Actions should never be performed withing use effect and only then when an action occured